go-controller/pkg/ovn/libovsdbops/nat.go

```diff
@@ -0,0 +1,100 @@
+package libovsdbops
+
+import (
+	"fmt"
+	"net"
+
+	libovsdbclient "github.com/ovn-org/libovsdb/client"
+	"github.com/ovn-org/libovsdb/model"
+	libovsdb "github.com/ovn-org/libovsdb/ovsdb"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/nbdb"
+)
+
+func buildRouterNAT(
+	natType nbdb.NATType,
+	externalIP net.IP,
+	logicalIP net.IP,
+	logicalPort string,
+	externalMac string,
+	) *nbdb.NAT {
+	return &nbdb.NAT{
+		Type:            natType,
+		ExternalIP:      externalIP.String(),
+		LogicalIP:       logicalIP.String(),
+		LogicalPort:     []string{logicalPort},
+		ExternalMAC:     []string{externalMac},
+		Options:         map[string]string{"stateless": "false"},
+	}
+}
+
+func CreateOrUpdateLogicalRouterNAT(nbClient libovsdbclient.Client, routerName string,
+	natType nbdb.NATType, externalIP net.IP, logicalIP net.IP, logicalPort string,
+	externalMac string) error {
+
+	ops := []libovsdb.Operation{}
+
+	routers := []nbdb.LogicalRouter{}
+	err := nbClient.WhereCache(func(item *nbdb.LogicalRouter) bool {
+		return item.Name == routerName
+	}).List(&routers)
+	if err != nil {
+		return fmt.Errorf("error finding logical router %s: %v", routerName, err)
+	}
+	if len(routers) == 0 {
+		return fmt.Errorf("found no logical routers with name %s", routerName)
+	}
+	if len(routers) > 1 {
+		return fmt.Errorf("unexpectedly found multiple logical routers: %+v", routers)
+	}
+	router := &routers[0]
+
+	nats := []nbdb.NAT{}
+	err = nbClient.WhereCache(func(item *nbdb.NAT) bool {
+		return item.Type == natType && item.ExternalIP == externalIP.String()
```

Comment:
Should logical ip be taken into account as well? Also I guess I would place the general logic of finding a NAT in a findNat method as well.

Reply:
Looking at what [the original code using] ovn-nbctl code does, it only "looks" for nat type and external ip. I think that logical ip is something can can be set via the update below, and there cannot be multiple rows for the sme external ip.

Ack, will move the logic to "findNAT" like tou are suggesting.

Reply:
Ok, I got a different impression from https://www.ovn.org/support/dist-docs/ovn-nb.5.html:

```
It  is  an error if a NAT already exists with the same values of
router, type, external_ip, and logical_ip, unless --may-exist is
specified.
```

Reply:
> Ok, I got a different impression from https://www.ovn.org/support/dist-docs/ovn-nb.5.html:
> 
> ```
> It  is  an error if a NAT already exists with the same values of
> router, type, external_ip, and logical_ip, unless --may-exist is
> specified.
> ```

The doc seems weird indeed! But there can't be multiple NATs with same type+external:
```
$ ovn-nbctl lr-add router
$ ovn-nbctl lr-nat-add router dnat_and_snat 1.1.1.1 2.2.2.2
$ ovn-nbctl lr-nat-add router dnat_and_snat 1.1.1.1 3.3.3.3
ovn-nbctl: a NAT with this type (dnat_and_snat) and external_ip (1.1.1.1) already exists
```